compiler/vm/compilerbridge.nim

```diff
@@ -0,0 +1,476 @@
+  addInNimDebugUtils(graph.config, "setupGlobalCtx")
+  if graph.vm.isNil:
+    let
+      ctx = newCtx(module, graph.cache, graph, idgen)
+      disallowDangerous =
+        defined(nimsuggest) or graph.config.cmd == cmdCheck or
+        vmopsDanger notin ctx.config.features
+
+    ctx.codegenInOut.flags = {cgfAllowMeta}
+    registerAdditionalOps(ctx[], disallowDangerous)
+
+    graph.vm = ctx
+  else:
+    let c = PCtx(graph.vm)
+    refresh(c[], module, idgen)
+
+proc evalConstExprAux(module: PSym; idgen: IdGenerator;
+                      g: ModuleGraph; prc: PSym, n: PNode,
+                      mode: TEvalMode): PNode =
+  addInNimDebugUtils(g.config, "evalConstExprAux", prc, n, result)
+  #if g.config.errorCounter > 0: return n
+  let n = transformExpr(g, idgen, module, n)
+  setupGlobalCtx(module, g, idgen)
+  let c = PCtx g.vm
+  let oldMode = c.mode
+  c.mode = mode
+  let requiresValue = mode!=emStaticStmt
+  let (start, regCount) = genExpr(c[], n, requiresValue).returnOnErr(c.config, n)
+
+  if c.code[start].opcode == opcEof: return newNodeI(nkEmpty, n.info)
+  assert c.code[start].opcode != opcEof
+  when defined(nimVMDebugGenerate):
+    c.config.localReport():
+      initVmCodeListingReport(c[], prc, n)
+
+  var tos = TStackFrame(prc: prc, comesFrom: 0, next: -1)
+  tos.slots.newSeq(regCount)
+  #for i in 0..<regCount: tos.slots[i] = newNode(nkEmpty)
+  let cb =
+    if requiresValue:
+      mkCallback(c, r): c.regToNode(r, n.typ, n.info)
+    else:
+      mkCallback(c, r): newNodeI(nkEmpty, n.info)
+
+  result = execute(c[], start, tos, cb).unpackResult(c.config, n)
+
+  c.mode = oldMode
+
+proc evalConstExpr*(module: PSym; idgen: IdGenerator; g: ModuleGraph; e: PNode): PNode =
+  result = evalConstExprAux(module, idgen, g, nil, e, emConst)
+
+proc evalStaticExpr*(module: PSym; idgen: IdGenerator; g: ModuleGraph; e: PNode, prc: PSym): PNode =
+  result = evalConstExprAux(module, idgen, g, prc, e, emStaticExpr)
+
+proc evalStaticStmt*(module: PSym; idgen: IdGenerator; g: ModuleGraph; e: PNode, prc: PSym) =
+  let r = evalConstExprAux(module, idgen, g, prc, e, emStaticStmt)
+  # TODO: the node needs to be returned to the caller instead
+  reportIfError(g.config, r)
+
+proc setupCompileTimeVar*(module: PSym; idgen: IdGenerator; g: ModuleGraph; n: PNode) =
+  let r = evalConstExprAux(module, idgen, g, nil, n, emStaticStmt)
+  # TODO: the node needs to be returned to the caller instead
+  reportIfError(g.config, r)
+
+proc setupMacroParam(reg: var TFullReg, c: var TCtx, x: PNode, typ: PType) =
+  case typ.kind
+  of tyStatic:
+    putIntoReg(reg, c, x, typ)
+  else:
+    var n = x
+    if n.kind in {nkHiddenSubConv, nkHiddenStdConv}: n = n[1]
+    # TODO: is anyone on the callsite dependent on this modifiction of `x`?
+    n.typ = x.typ
+    reg = TFullReg(kind: rkNimNode, nimNode: n)
+
+# XXX: why is this routine located here? It not related to the VM at all
```

Comment:
Could put in AST query?

Aside: got this far in my code review

Reply:
Done. The comment in question was new, but the rest of the code is for the largest part just copied over from `vm`.